File: top-k-frequent-elements.cpp
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> count;
		for (int num: nums) {
			if (count.find(num) == count.end()) {
				count[num] = 1;
			} else {
				count[num] += 1;
			}
		}

		vector<vector<int>> freq;
		for (int i=0; i<=nums.size(); ++i) {
			freq.push_back(vector<int>());
		}
		for (auto item: count) {
			freq[item.second].push_back(item.first);
		}

		vector<int> ans;
		for (int i=nums.size(); i>=0; --i) {
			for (int num: freq[i]) {
				ans.push_back(num);
				if (ans.size() == k) return ans;
			}
		}
		return ans;
    }
};

File: top-k-frequent-elements.py
from typing import List

class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        count = {}
        freq = [[] for _ in range(len(nums)+1)]

        for n in nums:
            count[n] = 1 + count.get(n, 0)

        for num, cnt in count.items():
            freq[cnt].append(num)

        ans = []
        for num_list in reversed(freq):
            for num in num_list:
                ans.append(num)
                if len(ans) == k: return ans
        return ans
